iothub/service/src/Messaging/FileUploadNotificationProcessorClient.cs

```diff
@@ -0,0 +1,174 @@
+    /// </summary>
+    /// <seealso href="https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-file-upload#service-file-upload-notifications"/>.
+    public class FileUploadNotificationProcessorClient : IDisposable
+    {
+        private readonly string _hostName;
+        private readonly IotHubConnectionProperties _credentialProvider;
+        private readonly IotHubConnection _connection;
+
+        /// <summary>
+        /// The callback to be executed each time file upload notification is received from the service.
+        /// </summary>
+        /// <remarks>
+        /// May not be null.
+        /// </remarks>
+        public Func<FileNotification, DeliveryAcknowledgement> _fileNotificationProcessor;
+
+        /// <summary>
+        /// The callback to be executed when the connection is lost.
+        /// </summary>
+        /// <remarks>
+        /// May not be null.
+        /// </remarks>
+        public Action<Exception> _errorProcessor;
+
+        /// <summary>
+        /// Creates an instance of this class. Provided for unit testing purposes only.
+        /// </summary>
+        protected FileUploadNotificationProcessorClient()
+        {
+        }
+
+        internal FileUploadNotificationProcessorClient(
+            string hostName,
+            IotHubConnectionProperties credentialProvider,
+            IotHubServiceClientOptions options)
+        {
+            _hostName = hostName;
+            _credentialProvider = credentialProvider;
+            _connection = new IotHubConnection(credentialProvider, options.UseWebSocketOnly, options.TransportSettings, options);
+            FileNotificationReceiver = new AmqpFileNotificationReceiver(_connection);
+        }
+
+        /// <summary>
+        /// Gets the AmqpFileNotificationReceiver which can deliver notifications for file upload operations.
+        /// </summary>
+        internal AmqpFileNotificationReceiver FileNotificationReceiver { get; }
+
+        /// <summary>
+        /// Open the AmqpFileNotificationReceiver instance.
+        /// </summary>
+        public virtual async Task OpenAsync()
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, $"Opening AmqpFileNotificationReceiver", nameof(OpenAsync));
+
+            try
+            {
+                await FileNotificationReceiver.OpenAsync().ConfigureAwait(false);
+                ReceivingAmqpLink receivingLink = await FileNotificationReceiver.FaultTolerantReceivingLink.GetReceivingLinkAsync().ConfigureAwait(false);
+                receivingLink.RegisterMessageListener(OnNotificationMessageReceivedAsync);
+            }
+            catch (Exception ex)
+            {
+                if (Logging.IsEnabled)
+                    Logging.Error(this, $"{nameof(OpenAsync)} threw an exception: {ex}", nameof(OpenAsync));
+                if (ex is IotHubException || ex is IOException)
+                {
+                    _errorProcessor?.Invoke(ex);
```

Comment:
There's no need to feed the error processor with this error when we throw it immediately afterwards. This error processor is only for surfacing connection level errors that occur in threads that the user does not maintain. Since the user maintains the thread that called ```OpenAsync()```, there is no need to feed this error to the error processor.